### ISSUE TITLE
[BUG] DUDE! Where's my WP_Post???

### ISSUE THREAD

**Issue by tn3rb** [timber/timber#2987]
### Expected Behavior

Hi I'm the lead dev for [Event Espresso](https://github.com/eventespresso/event-espresso-core) and we just received a support request from a user who is experiencing the following error:

```php
Fatal error: Uncaught Error: EE_CPT_Event_Strategy::the_post(): 
Argument #1 ($post) must be of type WP_Post, Timber\Post given
```

which is occurring because our plugin hooks into the following:

```php
/**
 * Fires once the post data has been set up.
 *
 * @since 2.8.0
 * @since 4.1.0 Introduced `$query` parameter.
 *
 * @param WP_Post  $post  The Post object (passed by reference).
 * @param WP_Query $query The current Query object (passed by reference).
 */
do_action_ref_array( 'the_post', array( &$post, &$this ) );
```

which is _supposed_ to provide a `WP_Post` object (passed by reference) for the first argument, but instead our callback is receiving an instance of `Timber\Post`. As well, we are ONLY hooking into `the_post` after it has already been established that the current post post_type is one of Event Espresso's custom post types.

We are slowly trying to modernize our codebase and take advantage of modern PHP functionality which includes utilizing Type declarations for function arguments. This of course will only work if other plugins and themes adhere to the expected argument types. 

Is our mutual user doing something wrong or does Timber actually replace the `WP_Post` object with a completely different object that is NOT compatible with `WP_Post`?!?!?!

I'm really REALLY hoping that this user has done something silly and Timber is NOT actually replacing the `WP_Post` object returned from 'the_post' action, which is wrong on so many levels. PLEASE tell me I'm not crazy!!!

### TLDR Expected Behaviour: 
To be able to use WordPress core functionality without having other plugins or themes replace expected arguments with incompatible data types.

### Actual behavior

```php
Fatal error: Uncaught Error: EE_CPT_Event_Strategy::the_post(): 
Argument #1 ($post) must be of type WP_Post, Timber\Post given
```

### Steps to reproduce behavior

1. add the following to any site using a Timber based theme:
   ```php
    add_action('the_post', 'dude_wheres_my_wp_post');
    function dude_wheres_my_wp_post( WP_Post $post ) 
    {
        echo "I'm supposed to be a WordPress Post object!";
    }
   ```
3. check out the completely unexpected error
...

### What version of Timber are you using?

Timber 2+

### What version of WordPress are you using?

WordPress 6.5

### What version of PHP are you using?

PHP 7.4+

### How did you install Timber?

None

**Comment by gchtr** [comment]
Timber hooks into `the_post` action by calling `setup_postdata()`, but we deliberately pass in a `WP_Post` object and not a `Timber\Post` instance. So I assume you’re not crazy :).

https://github.com/timber/timber/blob/7384e767d6cc208e5527674ea83ec25c6d21a17b/src/Post.php#L312-L313

I tried to reproduce the issue to see whether we missed something, but I can’t find an issue with Timber core itself. So I suspect that there’s something else going on.

It could be the theme or a plugin that adds a hook for `the_post` as well.

Because Timber v1 hooked into `the_post` [every time a `Timber\Post` was instantiated](https://github.com/timber/timber/blob/f8f2e5a9f03eb67b7c20820e60aa8618f59aaf1d/lib/Post.php#L612), some developers might have applied fixes themselves. We resolved this issue in Timber 2.0 so that `the_post` action is only called when looping over posts. On singular templates, `$post->setup()` has to be called manually. So any custom fix for this might not be needed anymore.

Some relevant documentation:

- https://timber.github.io/docs/v2/upgrade-guides/2.0/#better-compatibility-with-plugins
- https://timber.github.io/docs/v2/upgrade-guides/2.0/#template-context

**Comment by tn3rb** [comment]
> I assume you’re not crazy :)

kk thnx, although that's not what the voices are saying!  j/k (mental health is not a joke!)

ok we're going to assume that the user has done something funny in their template then. 
I'll pass those documentation links along to them.

Thank you so much for taking the time to look into this, it's greatly appreciated!

**Comment by Levdbas** [comment]
Good luck @tn3rb , and if the customer needs help, you can redirect him to this issue and we will gladly reopen.
